Schema Coordinates: add spec edit

magicmark opened the pull request:
👋

I've pulled out the proposed spec edits for Schema Coordinates (issue: https://github.com/graphql/graphql-spec/issues/735) into this PR.

(The RFC now lives in the original PR: https://github.com/graphql/graphql-spec/pull/746/files)

@leebyron you mentioned it might be possible to simplify the grammar - I played around a bit since the original PR, but I think I need some help here. What did you have in mind?

As suggested, tagging @eapache @mjmahone as additional reviewers as we had some good conversation in last WG meeting about this.

Thanks!

magicmark commented:
Oh also, here's another option I had for the examples table, was thinking it might be good to explicitly show all the different kinds of schema you can select:

<img width="857" alt="Screen Shot 2020-11-17 at 10 38 01 PM" src="https://user-images.githubusercontent.com/1590756/99497426-c5dea480-292a-11eb-9c9e-bee5b60902d4.png">

felt slightly overly verbose to me, but curious to see if anyone prefers this

benjie commented on spec/Appendix C -- Schema Coordinates.md:
```diff
@@ -0,0 +1,93 @@
+  - UnionName
+
+TypeDefinitionName:
+  - ObjectTypeName
+  - InterfaceTypeName
```
Don't forget input types, also these names aren't defined.

Since this is the grammar, I think it's actually more like:

```
SchemaCoordinate:
  - Name FieldSpecifier?
  - @ Name ArgumentSpecifier?

FieldSpecifier:
  - . Name ArgumentSpecifier?

ArgumentSpecifier:
  - ( ArgumentName : )
```

or maybe:

```
SchemaCoordinate:
  - TypeCoordinate
  - FieldCoordinate
  - FieldArgumentCoordinate
  - DirectiveCoordinate
  - DirectiveArgumentCoordinate

TypeCoordinate: Name

FieldCoordinate: Name . Name

FieldArgumentCoordinate: Name . Name ArgumentSpecifier

DirectiveCoordinate: @ Name

DirectiveArgumentCoordinate @ Name ArgumentSpecifier

ArgumentSpecifier: ( ArgumentName : )
```

I'm not happy using 'FieldCoordinate' to refer to enum values too, but since they use the same syntax they're seen as the same thing by the grammar. (Perhaps this is a good reason to use `EnumType->ENUM_VALUE` or similar to refer to them?)

leebyron replied:
Great clarification @benjie. I like the descriptiveness of the bottom one, but the top one is definitely more compact.

> I'm not happy using 'FieldCoordinate' 

`FieldOrEnumCoordinate` might make that less awkward?

> `ArgumentSpecifier: ( ArgumentName : )`

Should probably be `ArgumentSpecifier: ( Name : )`

benjie replied:
Oops yeah!

magicmark replied:
resolving this in favor of this thread: https://github.com/graphql/graphql-spec/pull/794#discussion_r527081282

benjie commented on spec/Appendix C -- Schema Coordinates.md:
````diff
@@ -0,0 +1,93 @@
+type Query {
+    searchBusiness(name: String!, filter: SearchFilter): Business
+}
+```
````
This example lacks custom scalars and input objects that would be useful to round it out.

magicmark replied:
Hmm hadn't thought about custom scalars! To clarify, assuming `scalar DateTime` or something is defined, we want `DateTime` to be a valid schema coordinate?

I guess it makes sense that you may want to refer to a custom scalar (since they can have directives applied, or maybe you just want to track how often a custom scalar is being used)

I guess it raises the question - what about built in scalars? Is `String` a valid schema coordinate too lol?

(I think as an artifact, it would have to be - just presumably not a very useful artifact...)

I'll call it out as an example though.

Thanks!!!

benjie replied:
Absolutely, built in scalars should be included too :+1:

benjie reviewed (changes requested):
Thanks for your hard work on this!

leebyron commented on spec/Appendix C -- Schema Coordinates.md:
```diff
@@ -0,0 +1,93 @@
+
+EnumValueSpecifier :
+  - . EnumValue
+
```
Also, it would be great to have a defined semantics to resolve something from a schema coordinate.

An example of this in the spec already are string semantics - https://spec.graphql.org/draft/#sec-String-Value.Semantics

This would avoid needing to completely rely on examples to illustrate all possible use cases

leebyron replied:
@magicmark and @benjie - folding your ideas together and adding semantics, how do you feel about this:

```
SchemaCoordinate:
  - Name
  - Name . Name
  - Name . Name ( Name : )
  - @ Name
  - @ Name ( Name : )

**Semantics**

SchemaCoordinate: Name
  1. Let {typeName} be the value of the first {Name}.
  2. Return the type in the schema named {typeName}.

SchemaCoordinate: Name . Name
  1. Let {typeName} be the value of the first {Name}.
  2. Let {type} be the type in the schema named {typeName}.
  3. Assert: {type} must exist.
  4. If {type} is an Enum type:
    1. Let {enumName} be the value of the second {Name}.
    2. Return the enum value of {type} named {enumName}.
  5. Otherwise if {type} is an Input Object type:
    1. Let {inputFieldName} be the value of the second {Name}.
    2. Return the input field of {type} named {inputFieldName}.
  6. Otherwise {type} must be an Object or Interface type:
    1. Let {fieldName} be the value of the second {Name}.
    2. Return the field of {type} named {fieldName}.
  
SchemaCoordinate: Name . Name ( Name : )
  1. Let {typeName} be the value of the first {Name}.
  2. Let {type} be the type in the schema named {typeName}.
  3. Assert: {type} must be an Object or Interface type.
  4. Let {fieldName} be the value of the second {Name}.
  5. Let {field} be the field of {type} named {fieldName}.
  6. Assert: {field} must exist.
  7. Let {argumentName} be the value of the third {Name}.
  8. Return the argument of {field} named {argumentName}.

SchemaCoordinate: @ Name
  1. Let {directiveName} be the value of the first {Name}.
  2. Return the directive in the schema named {directiveName}.

SchemaCoordinate: @ Name ( Name : )
  1. Let {directiveName} be the value of the first {Name}.
  2. Let {directive} be the directive in the schema named {directiveName}.
  3. Assert: {directive} must exist.
  7. Let {argumentName} be the value of the second {Name}.
  8. Return the argument of {directive} named {argumentName}.
```

benjie replied:
>   5. Otherwise {type} must be an Object or Interface type:

Maybe 
> 5. Otherwise:
>   1. Assert {type} is an Object or Interface type.
>   2. Let {fieldName} be...

LGTM :+1:

magicmark replied:
wow!

Ok this totally makes sense. I think I understand how to write the grammar a bit better now - I was trying to define the tokens by awkwardly encoding it with the naming of the token itself (e.g. "InterfaceTypeName").

This + the semantics section is clearly better!

Thank you so much @benjie and @leebyron for your work on this! <3

leebyron commented on spec/Appendix C -- Schema Coordinates.md:
```diff
@@ -0,0 +1,93 @@
+# Appendix: Schema Coordinates
```
Rather than an additional Appendix, might this reasonably fit as section 3.3.3 after Schema Extensions?

magicmark replied:
ack - had it as a separate thing to err on the side of not wanting to confuse/interfere with the actual runtime language syntax - but makes sense to me! moved as suggested.

magicmark commented on spec/Section 3 -- Type System.md:
```diff
@@ -232,6 +232,152 @@ Schema extensions have the potential to be invalid if incorrectly defined.
+
+SchemaCoordinate: Name
+  1. Let {typeName} be the value of the first {Name}.
+  2. Return the type in the schema named {typeName}.
```
Assuming that "type in the schema" applies to scalar types too? Is this worth clarifying?

benjie replied:
I don't think this needs clarification.

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -232,6 +232,152 @@ Schema extensions have the potential to be invalid if incorrectly defined.
+  1. Let {typeName} be the value of the first {Name}.
+  2. Let {type} be the type in the schema named {typeName}.
+  3. If {type} is an Enum type:
+    1. Let {enumName} be the value of the second {Name}.
+    2. Return the enum value of {type} named {enumName}.
```
I wonder if this should be `enumValueName` rather than `enumName` (which sounds to me like the type name).

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -232,6 +232,152 @@ Schema extensions have the potential to be invalid if incorrectly defined.
+  4. Otherwise if {type} is an Input Object type:
+    1. Let {inputFieldName} be the value of the second {Name}.
+    2. Return the input field of {type} named {inputFieldName}.
+  5. Otherwise {type} must be an Object or Interface type:
+    1. Let {fieldName} be the value of the second {Name}.
+    2. Return the field of {type} named {fieldName}.
```
I think making the assertion explicit is more in keeping with the rest of the algorithm:

```suggestion
  5. Otherwise:
    1. Assert: {type} must be an Object or Interface type.
    2. Let {fieldName} be the value of the second {Name}.
    3. Return the field of {type} named {fieldName}.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -232,6 +232,152 @@ Schema extensions have the potential to be invalid if incorrectly defined.
+| `Address`                      | `Address` interface                                                 |
+| `Address.city`                 | `city` field on the `Address` interface                             |
+| `ReviewInput`                  | `ReviewInput` input object type                                     |
+| `ReviewInput.author`           | `author` field on the `ReviewInput` input object type               |
```
```suggestion
| `ReviewInput.author`           | `author` input field on the `ReviewInput` input object type         |
```

benjie reviewed (approved):
Looks good to me; a few optional tweaks suggested.

magicmark commented on spec/Section 3 -- Type System.md:
```diff
@@ -232,6 +232,152 @@ Schema extensions have the potential to be invalid if incorrectly defined.
 
+### Schema Coordinates
+
+Schema Coordinates are human readable strings that uniquely identify an element defined in a GraphQL Schema.
```
I've also been using "member" (unintentionally) interchangeably, which I've heard others use in this context.

"element" feels more computery to me, but happy to go with whatever the established terminology is :D

leebyron replied:
I don't think we have an existing word for this concept, but if we're worried about this being unclear we could just be more verbose:

> Schema Coordinates are human readable strings that uniquely identify a specific type, field, argument, enum value, or directive defined in a GraphQL Schema.

leebyron commented on spec/Section 3 -- Type System.md:
```diff
@@ -232,6 +232,153 @@ Schema extensions have the potential to be invalid if incorrectly defined.
+
+**Definition**
+
+SchemaCoordinate:
```
I think we might want a `Note:` below this definition making it clear that `SchemaCoordinate` is not part of a `GraphQLDocument` and mentions the expected use case

magicmark commented on spec/Section 3 -- Type System.md:
```diff
@@ -232,6 +232,163 @@ Schema extensions have the potential to be invalid if incorrectly defined.
+| `DateTime`                     | Custom `DateTime` scalar type                                       |
+| `String`                       | Built-in `String` scalar type                                       |
+
+Schema Coordinates are always unique. Each type, field, argument, enum value, or
+directive may be referenced by exactly one possible Schema Coordinate.
```
feels like something worth highlighting and being clear about as a guarantee in the spec

magicmark commented on spec/Section 3 -- Type System.md:
````diff
@@ -232,6 +232,163 @@ Schema extensions have the potential to be invalid if incorrectly defined.
+Schema Coordinates are always unique. Each type, field, argument, enum value, or
+directive may be referenced by exactly one possible Schema Coordinate.
+
+For example, the following is *not* a valid Schema Coordinate:
+
+```graphql counter-example
+Entity.Business
+```
+
+In this counter example, both `Entity.Business` and `Business` would refer to
+the `Business` type. Since it would be ambiguous what the "primary key" should
+be in an application that uses Schema Coordinates to reference types, this is
+not allowed.
````
is this waffle? thought it might be helpful to have a little explanation/demonstration of this property

magicmark replied:
Could just be

```diff
- Since it would be ambiguous what the "primary key" should be in an application that uses Schema Coordinates to reference types, this is not allowed.
+ Such ambiguity is disallowed.
```

(More terse, but doesn't walk the reader through _why_ this is bad)

benjie replied:
> In this counter example, `Entity.Business` is redundant since `Business` already uniquely identifies the `Business` type. Such redundancy is disallowed by this spec - every type, field, field argument, enum value, directive, and directive argument has exactly one canonical Schema Coordinate.

magicmark replied:
👍 thanks benjie, stealing this wording

mcohen75 commented:
At Indeed we log usage of schema elements to 1) support deprecation and 2) understand usage from a product management perspective (i.e. is this API we built adding value?).

We defined our own coordinate system that is not quite as complete as this spec. A nice benefit of formalizing a coordinate system is that shared tools can be built to solve these and other use cases.

👍 👍 👍

leebyron commented:
I made some fairly significant edits to the prose and examples section in an attempt to simplify. @magicmark please take a look and let me know what you think.

leebyron commented:
IIRC the only thing holding this from Approval stage is a landed GraphQL.js PR?

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2067,3 +2067,114 @@ to the relevant IETF specification.
+*schema element* within a GraphQL Schema.
+
+:: A *schema element* is a specific instance of a named type, type field,
+input field, enum value, field argument, directive, or directive argument.
```
Great use of the new spec-md definitions feature :+1:

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2067,3 +2067,114 @@ to the relevant IETF specification.
+coordinates which refer to built-in schema elements.
+
+Note: A {SchemaCoordinate} is not a definition within a GraphQL {Document}, but
+a separate stand-alone grammar, intended to be used by tools to reference types,
```
I don't think it's necessary to hyphenate `stand-alone` these days:

```suggestion
a separate standalone grammar, intended to be used by tools to reference types,
```

leebyron replied:
TIL https://brians.wsu.edu/2016/05/31/standalone-stand-alone/

benjie reviewed (approved):
LGTM :+1:

benjie reviewed (approved):
Good catch on the Punctuator 👍

leebyron commented:
I think this is good to go again. I've just added the GraphQL.js implementation which greatly helped in completeness checking this spec change

leebyron commented:
Looks like the negative lookahead for `.` isn't rendering exactly right. I'll need to do a fix on spec-md before merging

benjie commented on spec/Appendix B -- Grammar Summary.md:
```diff
@@ -45,7 +45,7 @@ Token ::
   - StringValue
 
-Punctuator :: one of ! $ & ( ) ... : = @ [ ] { | } . [lookahead != .]
+Punctuator :: one of ! $ & ( ) ... : = @ [ ] { | } . [lookahead != {`.`, Digit}]
```
Good catch for decimals, but does the lookahead apply to the entire list or only the `.`? If the entire list, this currently forbids `foo: Int = 9` I think?

leebyron commented:
It’s meant to only apply to the dot

On Fri, Apr 16, 2021 at 8:34 AM Benjie Gillam ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In spec/Appendix B -- Grammar Summary.md
> <https://github.com/graphql/graphql-spec/pull/794#discussion_r614931374>:
>
> > @@ -45,7 +45,7 @@ Token ::
>    - FloatValue
>    - StringValue
>
> -Punctuator :: one of ! $ & ( ) ... : = @ [ ] { | } . [lookahead != .]
> +Punctuator :: one of ! $ & ( ) ... : = @ [ ] { | } . [lookahead != {`.`, Digit}]
>
> Good catch for decimals, but does the lookahead apply to the entire list
> or only the .? If the entire list, this currently forbids foo: Int = 9 I
> think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/graphql/graphql-spec/pull/794#pullrequestreview-637833405>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAMHUXJWXYQQ4JQU6W5L5DTJBKKHANCNFSM4TZSM7VA>
> .
>
-- 
Lee Byron

benjie commented:
> It’s meant to only apply to the dot

I had a quick read of the relevant parts of the spec and they didn't clarify that interpretation for me. There's doesn't seem to be any other examples of lookaheads applying to `one of` inline lists. We should try and remove the ambiguity here; I don't see any ambiguity in any of the other `lookahead` references.

_Aside: it seems that `spec-md` is adding the lookahead text via `::before` which means Chrome can't search for it in the productions; I had to look at the raw markdown._

leebyron commented:
I think I was trying to over-simplify by keeping this all in the `one of` and that's falling short on clarity. Here's an interesting existing case I'll try to model https://tc39.es/ecma262/#prod-Punctuator

leebyron commented:
At this point I think we're good to RFC 3 approve - I'd like to wait on a final OK from @magicmark and at least a first review from @IvanGoncharov on the reference impl changes (ideally a merge)

magicmark reviewed (commented):
Looking mighty fine!

PascalSenn commented on spec/Section 3 -- Type System.md:
```diff
@@ -2067,3 +2067,116 @@ to the relevant IETF specification.
+SchemaCoordinate : Name . Name
+  1. Let {typeName} be the value of the first {Name}.
+  2. Let {type} be the type in the {schema} named {typeName}.
+  3. If {type} is an Enum type:
+    1. Let {enumValueName} be the value of the second {Name}.
+    2. Return the enum value of {type} named {enumValueName}.
+  4. Otherwise if {type} is an Input Object type:
+    1. Let {inputFieldName} be the value of the second {Name}.
+    2. Return the input field of {type} named {inputFieldName}.
+  5. Otherwise:
+    1. Assert {type} must be an Object or Interface type.
+    2. Let {fieldName} be the value of the second {Name}.
+    3. Return the field of {type} named {fieldName}.
```
I think this section should also specify the behaviour for Union types.
At the moment the spec does not define the behaviour for `Union.UnionTypeMember` 

```suggestion
  3. If {type} is an Union type:
    1. Assert {type} must be an Enum, Input Object, Object or Interface type.
  4. If {type} is an Enum type:
    1. Let {enumValueName} be the value of the second {Name}.
    2. Return the enum value of {type} named {enumValueName}.
  5. Otherwise if {type} is an Input Object type:
    1. Let {inputFieldName} be the value of the second {Name}.
    2. Return the input field of {type} named {inputFieldName}.
  6. Otherwise:
    1. Assert {type} must be an Object or Interface type.
    2. Let {fieldName} be the value of the second {Name}.
    3. Return the field of {type} named {fieldName}.
```

magicmark replied:
> UnionTypeMember

The definition for this would live somewhere else in the schema, outside of the context of the union. You would write the schema coordinate to that instead.

The rule of thumb is: there should only be one way to reference a thing with Schema Coordinates

PascalSenn replied:
@magicmark  I understand, though the algorithm is still wrong.

PascalSenn replied:
@magicmark  Makes sense to me

magicmark replied:
@leebyron is it worth keeping in the clarification (or something similar) from a previous commit?

<img width="617" alt="Screen Shot 2021-12-31 at 4 42 48 AM" src="https://user-images.githubusercontent.com/1590756/147824058-8addb64b-76ef-4c82-b976-0dca94a96f44.png">

My gut says this may crop up again and cause some confusion - worth a counter example?

jord1e replied:
> @leebyron is it worth keeping in the clarification (or something similar) from a previous commit?
> 
> [magicmark/graphql-spec@`ac5318f`/spec/Section%203%20--%20Type%20System.md?plain=1#L391-L399](https://github.com/magicmark/graphql-spec/blob/ac5318f9a9f04d0f8dd03d3758d5ca2654f7f7d7/spec/Section%203%20--%20Type%20System.md?plain=1#L391-L399)
> 
> My gut says this may crop up again and cause some confusion - worth the counter example?

Just chiming in. From my perspective the counter example would be ideal (maybe even clarify that directives on types are also invalid?).

Maybe add a line on *why* union members are not valid schema coordinates, instead of solely stating that they are not?

```diff
- Note: You may not select members inside a union definition.
+ Note: Union members do not uniquely identify elements in the schema. Therefore the following counter example is *not* considered a valid Schema Coordinate:
```

...or something along those lines

magicmark replied:
fwiw this is all sort of encapsulated by the following line the spec:

> A *schema coordinate* is always unique. Each *schema element* may be referenced
> by exactly one possible schema coordinate.

++ for the explainer, I would maybe try and reuse this language, or hyperlink back to this source of truth somehow

schenkman commented:
Regarding input arguments, the example is shown as:

> Query.searchBusiness(name:)

What would it look like for more interesting sub-input-types on an argument? Example:

```graphql
Input BazInput {
  count: Int
  color: String
}
Input BarInput {
  baz_settings: BazInput
}
Input FooInput {
  bar_params: BarInput
}

type Query {
  foos(foo_params: FooInput) 
}
```

Would the above translate to the following coordinates?
```
Query.foos(foo_params:)
Query.foos(foo_params:bar_params)
Query.foos(foo_params:bar_params:baz_settings)
Query.foos(foo_params:bar_params:baz_settings:count:)
Query.foos(foo_params:bar_params:baz_settings:color:)
```

PascalSenn commented:
@schenkman 

This would be: 

Query.foos(foo_params:) => Query.foos(foo_params:)
Query.foos(foo_params:bar_params) => FooInput.bar_params
Query.foos(foo_params:bar_params:baz_settings:) => BarInput.baz_settings
Query.foos(foo_params:bar_params:baz_settings:count:) => BazInput.count

benjie commented:
Indeed, each schema entity should have exactly one schema coordinate. If you want to identify those kinds of paths in an operation (rather than the schema), then operation expressions might be what you’re after: https://github.com/graphql/graphql-wg/blob/main/rfcs/OperationExpressions.md#arguments

benjie commented:
@ddebrunner Schema Coordinates and Operation Expressions are two separate proposals with related syntax.

Every entity in a schema has exactly one schema coordinate, there is no schema coordinate `@A(foo_params:bar_params:)`, at least not as currently proposed.

The property `bar_params` on the type `FooInput` has the schema coordinate `FooInput.bar_params`, independent of whether it's accessed through a directive (`@A(foo_params:{bar_params:...})`) or a field (`myField(foo_params:{bar_params:...})`).

You can, however, use Operation Expressions to indicate the accessing of something in this manner, since Operation Expressions are not unique.

The comment you quote relates to Operation Expressions, which is a system that builds upon the Schema Coordinates syntax to talk about paths within operations (rather than the schema).

ddebrunner commented:
@benjie I'm not talking about operations, but schema elements. A type system directive is applied to a schema element and thus is also a schema element(?), so what are the coordinates to such arguments?

E.g.
```
input IO {
  a:IO2
}
input IO2 {
  b:Int
}
directive @x(v:IO)

type T {
   f:Int @x(v:{a:{b:3}})
}
```

Now if the value of `b` is invalid how would I report the coordinate of `b` in an error message.
The comments related to this RFC seem to imply something like:
```
T.f(v:a.b)
```

But I'm assuming that the use of type system directive`@x` is a schema element, and therefore so are its arguments.
Maybe I'm missing something?

benjie commented:
I don't think the _application_ of a directive is covered by this proposal, only the directive itself. At runtime your `@x` application on `T.f` no longer exists - e.g. if you introspect that schema or use `graphql.printSchema(schema)` you'll not see it - it's a build-time-only directive currently. (See #300 for more details.)

You could use operation expressions to indicate the path of this directive application, I guess it's more "document expressions" than "operation expressions" thinking about it.

ddebrunner commented:
Fair enough, but I hadn't assumed that the RFC was related to runtime, thus the fact that introspection doesn't return type system directives would seem to be irrelevant to the RFC.

cheapsteak commented:
Thoughts on being and to address field/type descriptions, field return types, and directives applied to fields/types? 🙏

(Our use case is to be able to address diffs in schema between a schema proposal and a schema check)

magicmark commented:
@cheapsteak do you have an example syntax for what you had in mind? If i'm understanding correctly, I think those are all derivable from the schema coordinate.

(Schema coordinates encode a unique location in the schema only. Adding in the above would be duplicative.)

cheapsteak commented:
@magicmark 

> @cheapsteak do you have an example syntax for what you had in mind? If i'm understanding correctly

yup for sure 🙏 

e.g. using Yelp's schema, something like `Business.id~description` to address the description of the field, or `Business~description` to address the description of the type, or `Business.id~returnType` to address the return type of the field

Our use case is allowing users to comment on a particular part of a schema, and while we can dangle them off of the type or field, it's not really the right place

magicmark commented:
> Our use case is allowing users to comment on a particular part of a schema, and while we can dangle them off of the type or field, it's not really the right place

Interesting. I understand now why the information encoded by schema coordinates isn't enough for you. Thanks!

I'm curious how common this use case is? (e.g. for something like github comments, comments are tied to lines of code rather than having a syntactic association)

I could see this making sense, but i'm hesitant to expand the scope and make any further substantive changes to this spec at this point

That said:

(1) there have been other proposals to expand the scope of schema coordinates, and companion specifications, so I think there's opportunity to extend this spec in future.
(2) if there are other compelling use cases and other folks agree it makes sense to do it in this spec, then a scope expansion could be warranted.

I'm planning to talk in the next WG meeting about this spec, maybe that's a good opportunity to chat about this if you want?

benjie commented:
Alternative syntax: `Business.id#description` inspired by Ruby/PHP/Elixir.

If we want to explore this, we should explore it further:

- `Animal#description`
- `Animal#members`
- `Animal#members#Cat` -- Not currently useful?
- `Animal#members#Cat#deprecationReason` -- Not currently possible to deprecate union members
- `Animal#members#Cat#description` -- Not currently possible: union members can't currently have descriptions. **Note**: NOT the same as `Cat#description`.
- `Cat#description`
- `Cat#deprecationReason`
- `Cat.lives#description`
- `Cat.lives#deprecationReason`
- `Cat.lives#type`
- `Cat.lives(includingSpent:)#type`
- `Cat.lives(includingSpent:)#description`
- `Cat.lives(includingSpent:)#deprecationReason`
- `MyEnum#description`
- `MyEnum#deprecationReason`
- `MyEnum.MY_VALUE#description`
- `MyEnum.MY_VALUE#deprecationReason`

Technically until we have #300 or similar we can't represent arbitrary directives on a position, but arguably we should consider that too:

- `MyEnum.MY_VALUE@myDirective{3}(arg:)` < the `arg` argument on the third `myDirective` directive applied to the MY_VALUE type on the `MyEnum` enum. :sweat_smile:

cheapsteak commented:
@benjie that all looks great! 🙏 although one nit with # is that it's not url safe and would need to be escaped

benjie commented:
Sadly same is true already of `:` - though most browsers can handle it unescaped, if you put it through `encodeURIComponent` it will be escaped:

```
> encodeURIComponent(':')
'%3A'
> new URLSearchParams({a: ':'}).toString()
'a=%3A'
```

magicmark commented:
thanks @cheapsteak!

I understand the use case now, makes sense.

at this point, things are pretty stable and we've already had many rounds of discussions on the syntax of this spec to reach alignment. it looks like we're near to being able to close this out.

so the question is, do we want to try and expand the scope of this spec and re-open it up, or follow it up with another spec?

my preference would be the latter - there are already discussions around expanding this spec in future (https://github.com/graphql/graphql-wg/blob/main/rfcs/OperationExpressions.md) so I think amending this spec as a separate RFC (which could be opened now) would make most sense to allow us to move forward on these proposals independently. (curious what @benjie thinks on this strategy)

benjie commented:
We wouldn't expand the scope of this RFC, **however** it's important to check that we're [preserving option value](https://github.com/graphql/graphql-spec/blob/main/CONTRIBUTING.md#guiding-principles) so we should explore related ideas to make sure we don't expect any conflicts.

In particular, I have concerns about the [Field Selection RFC](https://github.com/PascalSenn/composite-schemas-spec/blob/pse/adds-field-selection-rfc/rfcs/field-selection.md) which uses `.` for path traversal.

martinbonnin commented on spec/Section 3 -- Type System.md:
```diff
@@ -2164,3 +2164,128 @@ to the relevant IETF specification.
+:: A _schema coordinate_ is a human readable string that uniquely identifies a
+_schema element_ within a GraphQL Schema.
+
+:: A _schema element_ is a specific instance of a named type, field, input
+field, enum value, field argument, directive, or directive argument.
```
Nit: "instance" has very much a "runtime" connotation to me. Could we maybe get rid of it?

```suggestion
:: A _schema element_ can be a named type, a field, an input
a field, an enum value, a field argument, a directive, or a directive argument.
```

martinbonnin commented on spec/Section 3 -- Type System.md:
```diff
@@ -2164,3 +2164,128 @@ to the relevant IETF specification.
+
+A _schema coordinate_ may refer to either a defined or built-in _schema
+element_. For example, `String` and `@deprecated(reason:)` are both valid schema
+coordinates which refer to built-in schema elements. However it must not refer
+to a meta-field. For example, `Business.__typename` is _not_ a valid schema
+coordinate.
```
Is there a specific reason to forbid meta fields? I can definitely see myself using it. For an example to log introspection usages of my server.

magicmark replied:
Good question! I don't actually remember if there was a technical reason, or if was around limiting the spec to the intended application use cases 🤔 

This [blames](https://github.com/graphql/graphql-spec/commit/b65eb313af7a844d243473743881412cad049014) to @leebyron, so ping Lee to see if he remembers.

magicmark replied:
@leebyron any thoughts?

martinbonnin commented on spec/Section 3 -- Type System.md:
```diff
@@ -2164,3 +2164,128 @@ to the relevant IETF specification.
+
+SchemaCoordinate : Name
+
+1. Let {typeName} be the value of the first {Name}.
+2. Return the type in the {schema} named {typeName}.
+
```
Nit: should we assert that the type exists?

```suggestion
1. Let {typeName} be the value of the first {Name}.
2. Let {type} be the type in the {schema} named {typeName}.
3. Assert {type} exists.
4. Return {type}.

```

This is mainly for symmetry with the "argument coordinate" case below which asserts field existence and are therefore allowed to "fail".  Asserting that the type exists makes it explicit that a coordinate may "fail".

benjie replied:
I don't think we should do this. For example, if you wanted to ask the question "Does `Foo.bar(baz:)` exist in the schema, you might first check the schema coord `Foo` and make sure it's an object or interface, and then check `Foo.bar` exists, and finally access `Foo.bar(baz:)`. It should not be required that checking for `Foo` causes it to throw IMO - returning `null` seems not unreasonable. Throwing if you try to access `Foo.bar(baz:)` when `Foo.bar` doesn't exist does however seem reasonable.

martinbonnin replied:
Does it even make sense to have "throw" or return "null"? I mean this is all declarative, it should be possible to represent an invalid coordinate if people want (for counter examples or whatever...). 

If we were to actually implement a function I would probably use a verb:

```
GetAstNode(SchemaCoordinate):
...
```

Do we want graphql-js to have such functions?

benjie replied:
That's effectively what we're doing

> To refer to a _schema element_, a _schema coordinate_ must be interpreted in the context of a GraphQL {schema}.

I.e. a schema coordinate can exist, but if you want to refer to a schema element then you follow the algorithm.

martinbonnin replied:
Alright so what we're saying is:

```
throws if any component leading to the last component.
returns null if the last component of the coordinate is not found.
else return thee element.
```

If we're going that road, I would say we probably need to be explicit about it:

```
1. Let {typeName} be the value of the first {Name}.
2. If there is not type in the schema named {typeName} return null.
2. Else let {type} be the type in the {schema} named {typeName}.
4. return {type}.
```

magicmark replied:
> it should be possible to represent an invalid coordinate if people want (for counter examples or whatever...)

At this risk of going down a rabbit hole of over-explicitness -- the spec as written states "a _schema coordinate_ must be interpreted in the context of a GraphQL {schema}." 

We are encoding parsing and validating within the same set of instructions here.

If we wanted to explicitly allow for "representing an invalid coordinate" without someone tutting at you, we _could_ expose seperate parse() and validate() functions in the graphql-js implementation, which would mirror this spec edit

martinbonnin replied:
@magicmark I like this 👍 . It's also more consistent with the rest of the spec.

martinbonnin replied:
After reading the [graphql-js PR](https://github.com/graphql/graphql-js/pull/3044/files), I think it'd make sense to have `resolveTypeCoordinate(TypeCoordinate)`, `resolveMemberCoordinate(MemberCoordinate)`, etc...

I can look into this for the next working group if there is interest, just let me know.

martinbonnin commented on spec/Appendix B -- Grammar Summary.md:
```diff
@@ -43,7 +43,14 @@ Token ::
 - StringValue
 
-Punctuator :: one of ! $ & ( ) ... : = @ [ ] { | }
+Punctuator ::
+
+- DotPunctuator
+- OtherPunctuator
```
Out of curiosity, why do we even need that lexical rule?  Most of the parsers I've bumped into do not have a `Punctuator` token and just use the underlying character directly. `Punctuator` doesn't seem to be used by any grammar rule either?

benjie replied:
It's one of the `Token`s supported by the lexer; lex the input stream into tokens, then parse the tokens into an AST.

martinbonnin commented on spec/Section 3 -- Type System.md:
```diff
@@ -2164,3 +2164,128 @@ to the relevant IETF specification.
+:: A _schema coordinate_ is a human readable string that uniquely identifies a
+_schema element_ within a GraphQL Schema.
+
+:: A _schema element_ can be a named type, a field, an input
+a field, an enum value, a field argument, a directive, or a directive argument.
```
Woops, I put an `a` too many, apologies 😬 

```suggestion
:: A _schema element_ can be a named type, a field, an input 
field, an enum value, a field argument, a directive, or a directive argument.
```

benjie commented:
GraphQL.js PR is up to date again; let's get this merged.

https://github.com/graphql/graphql-js/pull/3044

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+
+Note: A {SchemaCoordinate} is not a definition within a GraphQL {Document}, but
+a separate standalone grammar, intended to be used by tools to reference types,
+fields, and other *schema element*s. Examples include: as references within
```
We never pluralise references. I know, its weird.

```suggestion
fields, and other _schema element_. Examples include: references within
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+context of a GraphQL {schema}.
+
+If the _schema element_ cannot be found, the resolve function will not yield a
+value (without raising an error). However, an error **will** be raised if any
```
```suggestion
value (without raising an error). However, an error will be raised if any
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+
+SchemaCoordinate : Name . Name
+
+1. Let {typeName} be the value of the first {Name}.
+2. Let {type} be the type in the {schema} named {typeName}.
+3. Assert {type} must exist.
+4. If {type} is an Enum type:
+5. Let {enumValueName} be the value of the second {Name}.
+6. Let {enumValue} be the enum value of {type} named {enumValueName}.
+7. If {enumValue} does not exist, return {null}.
+8. Return {enumValue}.
+9. Otherwise if {type} is an Input Object type:
+10. Let {inputFieldName} be the value of the second {Name}.
+11. Let {inputField} be the input field of {type} named {inputFieldName}.
+12. If {inputField} does not exist, return {null}.
+13. Return {inputField}.
+14. Otherwise:
+15. Assert {type} must be an Object or Interface type.
+16. Let {fieldName} be the value of the second {Name}.
+17. Let {field} be the field of {type} named {fieldName}.
+18. If {field} does not exist, return {null}.
+19. Return {field}.
```
Fixed the indentation. I also think these algorithms are a little verbose (could use 2 steps instead of 4, with second being e.g. `Return the enum value of {type} named {enumValueName}, or {null} if no such value exists.`

```suggestion
1. Let {typeName} be the value of the first {Name}.
2. Let {type} be the type in the {schema} named {typeName}.
3. Assert: {type} must exist.
4. If {type} is an Enum type:
   1. Let {enumValueName} be the value of the second {Name}.
   2. Let {enumValue} be the enum value of {type} named {enumValueName}.
   3. If {enumValue} does not exist, return {null}.
   4. Return {enumValue}.
5. Otherwise, if {type} is an Input Object type:
   1. Let {inputFieldName} be the value of the second {Name}.
   2. Let {inputField} be the input field of {type} named {inputFieldName}.
   3. If {inputField} does not exist, return {null}.
   4. Return {inputField}.
6. Otherwise:
   1. Assert {type} must be an Object or Interface type.
   2. Let {fieldName} be the value of the second {Name}.
   3. Let {field} be the field of {type} named {fieldName}.
   4. If {field} does not exist, return {null}.
   5. Return {field}.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+2. Let {directive} be the directive in the {schema} named {directiveName}.
+3. Assert {directive} must exist.
+4. Let {directiveArgumentName} be the value of the second {Name}.
+5. Let {directiveArgument} be the argument of {directive} named.
+   {directiveArgumentName}.
```
```suggestion
5. Let {directiveArgument} be the argument of {directive} named
   {directiveArgumentName}.
```

benjie reviewed (changes requested):
I thought we had moved to using ULs for all our algorithms, but seems that section 3 is still full of OLs!

Since these are minor syntactic fixes I'll go ahead and apply them myself

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+
+1. Let {typeName} be the value of the first {Name}.
+2. Let {type} be the type in the {schema} named {typeName}.
+3. Assert {type} must exist.
+4. Assert {type} must be an Object or Interface type.
```
```suggestion
3. Assert: {type} must exist.
4. Assert: {type} must be an Object or Interface type.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+4. Assert {type} must be an Object or Interface type.
+5. Let {fieldName} be the value of the second {Name}.
+6. Let {field} be the field of {type} named {fieldName}.
+7. Assert {field} must exist.
```
```suggestion
7. Assert: {field} must exist.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+SchemaCoordinate : Name
+
+1. Let {typeName} be the value of the first {Name}.
+2. Let {type} be the type in the {schema} named {typeName}.
+3. If {type} does not exist, return {null}.
+4. Return {type}.
```
```suggestion
2. Return the type in the {schema} named {typeName}, or {null} if no such type exists.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+3. Assert: {type} must exist.
+4. If {type} is an Enum type:
+   1. Let {enumValueName} be the value of the second {Name}.
+   2. Let {enumValue} be the enum value of {type} named {enumValueName}.
+   3. If {enumValue} does not exist, return {null}.
+   4. Return {enumValue}.
```
```suggestion
   2. Return the enum value of {type} named {enumValueName}, or {null} if no such value exists.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+   4. Return {enumValue}.
+5. Otherwise, if {type} is an Input Object type:
+   1. Let {inputFieldName} be the value of the second {Name}.
+   2. Let {inputField} be the input field of {type} named {inputFieldName}.
+   3. If {inputField} does not exist, return {null}.
+   4. Return {inputField}.
```
```suggestion
   2. Return the input field of {type} named {inputFieldName}, or {null} if no such input field exists.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+6. Otherwise:
+   1. Assert: {type} must be an Object or Interface type.
+   2. Let {fieldName} be the value of the second {Name}.
+   3. Let {field} be the field of {type} named {fieldName}.
+   4. If {field} does not exist, return {null}.
+   5. Return {field}.
```
```suggestion
   3. Return the field of {type} named {fieldName}, or {null} if no such field exists.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+6. Let {field} be the field of {type} named {fieldName}.
+7. Assert: {field} must exist.
+8. Let {fieldArgumentName} be the value of the third {Name}.
+9. Let {fieldArgument} be the argument of {field} named {fieldArgumentName}.
+10. If {fieldArgument} does not exist, return {null}.
+11. Return {fieldArgument}.
```
```suggestion
9. Return the argument of {field} named {fieldArgumentName}, or {null} if no such argument exists.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+SchemaCoordinate : @ Name
+
+1. Let {directiveName} be the value of the first {Name}.
+2. Let {directive} be the directive in the {schema} named {directiveName}.
+3. If {directive} does not exist, return {null}.
+4. Return {directive}.
```
```suggestion
2. Return the directive in the {schema} named {directiveName}, or {null} if no such directive exists.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,149 @@ to the relevant IETF specification.
+2. Let {directive} be the directive in the {schema} named {directiveName}.
+3. Assert: {directive} must exist.
+4. Let {directiveArgumentName} be the value of the second {Name}.
+5. Let {directiveArgument} be the argument of {directive} named
+   {directiveArgumentName}.
+6. If {directiveArgument} does not exist, return {null}.
+7. Return {directiveArgument}.
```
```suggestion
5. Return the argument of {directive} named
   {directiveArgumentName}, or {null} if no such argument exists.
```

benjie reviewed (approved):
I still approve of this addition :+1: 

There is no basis for this preference in the spec currently as far as I can tell, but these algorithms seem unnecessarily verbose... I've attached some suggestions based on this pattern:

```md
    - Let {groupForResponseName} be the list in {groupedFields} for
      {responseName}; if no such list exists, create it as an empty list.
```

martinbonnin commented on spec/Appendix B -- Grammar Summary.md:
```diff
@@ -412,3 +419,11 @@ TypeSystemDirectiveLocation : one of
 - `INPUT_OBJECT`
 - `INPUT_FIELD_DEFINITION`
+
+SchemaCoordinate :
+
+- Name
+- Name . Name
+- Name . Name ( Name : )
+- @ Name
+- @ Name ( Name : )
```
Should this also use the intermediates nodes?

```suggestion
SchemaCoordinate :

- TypeCoordinate
- MemberCoordinate
- ArgumentCoordinate
- DirectiveCoordinate
- DirectiveArgumentCoordinate

TypeCoordinate : Name

MemberCoordinate : Name . Name

ArgumentCoordinate : Name . Name ( Name : )

DirectiveCoordinate : @ Name

DirectiveArgumentCoordinate : @ Name ( Name : )
```

magicmark commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,150 @@ to the relevant IETF specification.
+
+MemberCoordinate : Name . Name
+
+1. Let {typeName} be the value of the first {Name}.
```
```suggestion
1. Let {typeName} be the value of the first {Name}.
2. Assert: {typeName} must not be a meta-field.
```

magicmark replied:
Teeing this up in anticipation that we decide on this for now.

Could also be phrased as `Assert: {typeName} must not begin with {__} (two underscores)`

Given this:

> All types and directives defined within a schema must not have a name which begins with "__" (two underscores), as this is used exclusively by GraphQL’s introspection system.

but former seems preferred for brevity?

(that is, assuming that all fields prefixed with `__` are meta-fields and vice versa. I think this is the case but not explicitly stated this is the same set afaict?)

magicmark commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,150 @@ to the relevant IETF specification.
+MemberCoordinate : Name . Name
+
+1. Let {typeName} be the value of the first {Name}.
+2. Let {type} be the type in the {schema} named {typeName}.
```
```suggestion
2. Let {type} be the type in the {schema} named {typeName}.
3. Assert: The value of the second {Name} does not begin with "__" (two underscores).
```

magicmark commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,150 @@ to the relevant IETF specification.
+1. Let {typeName} be the value of the first {Name}.
+2. Let {type} be the type in the {schema} named {typeName}.
+3. Assert: {type} must exist, and be an Object or Interface type.
+4. Let {fieldName} be the value of the second {Name}.
```
e.g. 
__type(name: String!): __Type

https://spec.graphql.org/October2021/#sec-Schema-Introspection

magicmark commented on spec/Section 3 -- Type System.md:
```diff
@@ -2168,3 +2168,150 @@ to the relevant IETF specification.
+
+TypeCoordinate : Name
+
+1. Let {typeName} be the value of {Name}.
```
```suggestion
1. Let {typeName} be the value of {Name}.
1. Assert: {typeName} must not begin with "__" (two underscores).
```

BoD commented on spec/Appendix B -- Grammar Summary.md:
```diff
@@ -412,3 +419,21 @@ TypeSystemDirectiveLocation : one of
+
+MemberCoordinate : Name . Name
+
+ArgumentCoordinate : Name . Name ( Name : )
```
Any reason why `type.field(arg:)` was preferred over `type.field(arg)`? 

Apologies if this has already been discussed before.

magicmark replied:
We went back and forth on this a few times when this was first proposed. Personally I don't care either way (slightly prefer `type.field(arg)` even) but `arg:` mirrors other languages (e.g.)
```suggestion
ArgumentCoordinate : Name . Name ( Name : )https://docs.swift.org/swift-book/documentation/the-swift-programming-language/functions/#:~:text=The%20body%20of%20the%20greet(person%3A)
```

benjie commented:
There has been discussion around whether having the coordinate for a field and an enum value be the same makes sense. I propose that we change the syntax used for enum values from `.` to `::`:

```
Type # Any named type
Type.field # A field (e.g. on an object, interface, or inputObject type)
Type.field(argName:) # An argument (on an object or interface type)
Type::value # an Enum value (on an enum type)
```

`::` is the syntax used to reference an enum value in [C++](https://en.cppreference.com/w/cpp/language/enum.html), [Rust](https://doc.rust-lang.org/reference/items/enumerations.html), [PHP](https://www.php.net/manual/en/language.enumerations.basics.php), [Perl](https://manpages.ubuntu.com/manpages/questing/man3/enum.3pm.html) and others, so it seems pretty natural.

I also think that doing this will make it more obvious for users reading a schema coordinate that it's an enum value being referenced rather than a field (and vice versa).

benjie commented:
I have raised a PR to implement these changes here: https://github.com/magicmark/graphql-spec/pull/1

leebyron commented on spec/Appendix B -- Grammar Summary.md:
```diff
@@ -412,3 +422,24 @@ TypeSystemDirectiveLocation : one of
+
+ArgumentCoordinate : Name . Name ( Name : )
+
+ValueCoordinate : Name :: Name
```
agree EnumValueCoordinate is more clear

magicmark replied:
Reverting to `MemberCoordinate`.

benjie commented:
Conclusion from the WG was to:

a) support meta fields
b) use the same syntax for fields and enum values.

Sorry @JoviDeCroock, you were outvoted! I think your concerns have merit and hopefully I represented them well, but equally what we're primarily doing here is encoding a pattern we already see in the ecosystem and apparently people are actively encoding enum values like this (who knew? I only ever see schema coordinates used for types, fields, arguments and directives!) so "member" it is! :shrug:

magicmark commented:
Updates June 6th 2025:

- I have reverted back to proposing MemberCoordinates (`Name.Name`) to refer to both object fields and enum values, per the WG discussion and reasoning above.
- meta-fields are no longer disallowed (@leebyron doesn't remember why we added this restriction in the first place)
- union members are still disallowed. I don't have a super strong preference for if we should allow that or not (I personally can't think of use cases for it), but either way, my strong personal preference would be to tackle this in a future spec proposal, should we decide to add support for that.

benjie reviewed (changes requested):
Really excited that we're getting close to this being ready to launch!

We still need to reach a consensus on the "meta fields" and "introspection types" question. I think the real issue comes down to "resolution" and I don't think you should be able to "resolve" a meta field or a member of the introspection schema (since these may be implemented in completely different ways to regular fields/types/etc) - and thus we should exclude them from _schema element_ explicitly, and let the rest of the logic flow from there. This doesn't make them "invalid" so much as "implementation defined behavior".

magicmark commented on spec/Section 3 -- Type System.md:
```diff
@@ -2227,3 +2227,160 @@ to the relevant IETF specification.
+
+A _schema coordinate_ is always unique. Each _schema element_ can be referenced
+by exactly one possible schema coordinate.
+
```
```suggestion

Note: To ensure uniqueness, {Ignored} tokens not valid inside a _schema coordinate_.
```

magicmark replied:
@benjie I suppose it would be most correct to add a `lookahead` rule to the grammars above, but that would end up being super super verbose - is there any shorthand for this?

magicmark replied:
In fact:

> Ignored tokens are [...] not referenced in syntactical grammar productions.
> Any amount of Ignored may appear before and after every lexical token

~ https://spec.graphql.org/October2021/#sec-Language.Source-Text.Ignored-Tokens

So we'd want to change this.... but only in the case of when parsing schema coordinates? Which would lead to a tricky change to the parse() function... 🤔 

(almost feels like at this point you'd be easier off with an entirely different and dedicated `parseSchemaCoordinate()` function)

magicmark replied:
One option of course it to fudge it and add a little context to our otherwise(?) context free grammar (lots of prior art in other languages) and add a special rule in the parser that says "if you're about to yield a schema coordinate, bail out if you encountered any {Ignored} tokens"

I'd maybe vote for that?

magicmark replied:
took a stab at implementation here: https://github.com/graphql/graphql-js/pull/4450

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2227,3 +2227,160 @@ to the relevant IETF specification.
+fields, and other _schema element_. Examples include: references within
+documentation to refer to types and fields in a schema, a lookup key that can be
+used in logging tools to track how often particular fields are queried in
+production.
```
How about this, @magicmark?

```suggestion
production.

**Lexical Analysis & Syntactic Parse of a Schema Coordinate**

The source text of a _schema coordinate_ is processed in the same way as that of
a GraphQL document, as laid out in the [Language](#sec-Language) section, with
the exception that {Ignored} tokens are disallowed. A schema coordinate must
therefore not contain whitespace, line terminators, comments, commas, or a _Byte
Order Mark_. This constraint ensures that schema coordinates are both
unambiguous and unique.
```

magicmark replied:
Added a "with the exception of _schema coordinate_" here to avoid a contradiction

https://github.com/graphql/graphql-spec/pull/794/commits/f4eec91893d26a65d75bdeba73a6a815190ab710

benjie commented:
@magicmark Here's my original diff that I drafted up locally:

```diff
diff --git a/spec/Section 2 -- Language.md b/spec/Section 2 -- Language.md
index e422d9a..7a6dc1f 100644
--- a/spec/Section 2 -- Language.md	
+++ b/spec/Section 2 -- Language.md	
@@ -113,6 +113,8 @@ Comments are {Ignored} like white space and may appear after any token, or
 before a {LineTerminator}, and have no significance to the semantic meaning of a
 GraphQL Document.
 
+Comments are forbidden within a _schema coordinate_.
+
 ### Insignificant Commas
 
 Comma :: ,
@@ -154,25 +156,27 @@ Ignored ::
 - Comma
 
 {Ignored} tokens are used to improve readability and provide separation between
-lexical tokens, but are otherwise insignificant and not referenced in
-syntactical grammar productions.
+lexical tokens in a GraphQL document, but are otherwise insignificant and not
+referenced in syntactical grammar productions. {Ignored} are forbidden within a
+_schema coordinate_.
 
-Any amount of {Ignored} may appear before and after every lexical token. No
-ignored regions of a source document are significant, however {SourceCharacter}
-which appear in {Ignored} may also appear within a lexical {Token} in a
-significant way, for example a {StringValue} may contain white space characters.
-No {Ignored} may appear _within_ a {Token}, for example no white space
-characters are permitted between the characters defining a {FloatValue}.
+Any amount of {Ignored} may appear before and after every lexical token in a
+GraphQL document. No ignored regions of a source document are significant,
+however {SourceCharacter} which appear in {Ignored} may also appear within a
+lexical {Token} in a significant way, for example a {StringValue} may contain
+white space characters. No {Ignored} may appear _within_ a {Token}, for example
+no white space characters are permitted between the characters defining a
+{FloatValue}.
 
 **Byte Order Mark**
 
 UnicodeBOM :: "Byte Order Mark (U+FEFF)"
 
-The _Byte Order Mark_ is a special Unicode code point which may appear at the
+:: The _Byte Order Mark_ is a special Unicode code point which may appear at the
 beginning of a file which programs may use to determine the fact that the text
 stream is Unicode, and what specific encoding has been used. As files are often
 concatenated, a _Byte Order Mark_ may appear before or after any lexical token
-and is {Ignored}.
+and is {Ignored}. {UnicodeBOM} is forbidden within a _schema coordinate_.
 
 ### Punctuators
```

Ultimately I decided that section 2 of the spec is about the GraphQL language (for GraphQL documents) which does _not_ specifically relate to schema coordinates, and so instead of talking about schema coordinates in section 2 before they are introduced, I would instead have schema coordinates reference section 2 and apply the exception on top. I'm not sure if this is the right decision - @leebyron would be able to better weigh in here - but if we are to comment in section 2 then here are the kind of changes that I would have made (i.e. adding "in a GraphQL document" to differentiate such from a "schema coordinate").

benjie commented on spec/Appendix B -- Grammar Summary.md:
```diff
@@ -413,3 +413,24 @@ TypeSystemDirectiveLocation : one of
 - `ENUM_VALUE`
 - `INPUT_OBJECT`
 - `INPUT_FIELD_DEFINITION`
+
```
This is currently under the title "Document Syntax"; we should use it's own header.

```suggestion

## Schema Coordinate Syntax

Note: In order to be unique, schema coordinates do not permit {Ignored} tokens.
```

benjie commented on spec/Appendix B -- Grammar Summary.md:
```diff
@@ -413,3 +413,24 @@ TypeSystemDirectiveLocation : one of
+DirectiveCoordinate : @ Name
+
+DirectiveArgumentCoordinate : @ Name ( Name : )
+
+Note: This grammar is unique. A _schema coordinate_ may not include any other
+tokens.
```
Moved note and clarified.

```suggestion
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2227,3 +2227,172 @@ to the relevant IETF specification.
+element_. For example, `String` and `@deprecated(reason:)` are both valid schema
+coordinates which refer to built-in schema elements.
+
+Note: Union members are not valid _schema coordinate_ as they reference existing
+types in the schema. This preserves the uniqueness property of a _schema
+coordinate_ as stated above.
```
```suggestion
Note: A union member references a type in the schema. A type in the schema is
identified by a {TypeCoordinate}. There is no schema coordinate which indicates
a union member; this preserves the uniqueness property of a _schema coordinate_
as stated above.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2227,3 +2227,172 @@ to the relevant IETF specification.
+
+**Lexical Analysis & Syntactic Parse of a Schema Coordinate**
+
+The source text of a _schema coordinate_ is processed in the same way as that of
+a GraphQL document, as laid out in the [Language](#sec-Language) section, with
+the exception that only the subset of tokens as described in the grammar above
+may be considered. A schema coordinate must therefore not contain whitespace,
+line terminators, comments, commas, or a _Byte Order Mark_. This constraint
+ensures that schema coordinates are both unambiguous and unique.
```
```suggestion
Similar to [GraphQL documents](#sec-Language), a schema coordinate is defined as
a syntactic grammar where terminal symbols are tokens (indivisible lexical
units). These tokens are defined in a lexical grammar which matches patterns of
source characters. The source text of a schema coordinate must be a sequence of
{SourceCharacter}. The character sequence must be described by a sequence of
{Token} lexical grammars. The lexical token sequence, must be described by a
single {SchemaCoordinate} syntactic grammar.

Unlike with GraphQL documents, {Ignored} lexical grammars are not permitted
within the character sequence; a schema coordinate must therefore not contain
whitespace, line terminators, comments, commas, or a _Byte Order Mark_. This
constraint ensures that schema coordinates are both unambiguous and unique.
```

benjie commented on spec/Section 3 -- Type System.md:
```diff
@@ -2227,3 +2227,172 @@ to the relevant IETF specification.
+
+DirectiveArgumentCoordinate : @ Name ( Name : )
+
+Note: A _schema coordinate_ is defined with its own grammar and may not include
+any other tokens.
```
```suggestion
Note: A _schema coordinate_ is defined with its own grammar that, unlike GraphQL
documents, excludes {Ignored} lexical grammars (see: [Lexical Analysis &
Syntactic Parse of a Schema Coordinate](
(#sec-Type-System.Schema-Coordinates.Lexical-Analysis-Syntactic-Parse-of-a-Schema-Coordinate)).
```
